### PR TITLE
[Quant][PT2E] Supported customized _EQUIVALENT_TYPES in Module Partition API

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98826
* #104245
* #104075
* __->__ #102516

**Summary**
`Module Partition API` can simplify the pattern match process in Quantization annotation. However, current implementation of 
`Module Partition API` has hardcoded `_EQUIVALENT_TYPES` https://github.com/pytorch/pytorch/blob/999bae0f54108ffc5b7cf2524a02a83901554b16/torch/ao/quantization/_pt2e/graph_utils.py#L13-L20. So, PyTorch Extension Libraries such as [intel-extension-for-pytorch](https://github.com/intel/intel-extension-for-pytorch) can't use `Module Partition API` with customized `_EQUIVALENT_TYPES` . In this PR, we plan to enable customized `_EQUIVALENT_TYPES` by pass in parameter. 

**Test Plan**
```
python -m pytest test_graph_utils.py -k test_customized_equivalet_types_dict
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Comment by kimishpatel** [comment]
I dont see any use of the added api in this stack. Can you add an example of why you need this change? If needed, we need to think of interface where it allows you to not just override but also extend the equivalent_types.
e.g.
```
def get_equivalent_types()
"""
defines a map from nn.Module equivalent nn.functional
"""
```

File: torch/ao/quantization/_pt2e/graph_utils.py [hunk 1]
```diff
@@ -42,17 +47,20 @@ def _partitions_sequential(partitions: List[SourcePartition]):
     return True
 
 
-def _get_matching_types(partition_type):
+def _get_matching_types(partition_type, equivalet_types_dict=None):
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply torch/ao/quantization/_pt2e/graph_utils.py:50]: Thanks and fix this and other spelling mistakes.

File: torch/ao/quantization/_pt2e/graph_utils.py [hunk 2]
```diff
@@ -42,17 +47,20 @@ def _partitions_sequential(partitions: List[SourcePartition]):
...
     matching_types = [partition_type]
-    if partition_type in _EQUIVALET_TYPES_DICT:
-        matching_types.extend(_EQUIVALET_TYPES_DICT[partition_type])
+    if partition_type in equivalet_types_dict:
+        matching_types.extend(equivalet_types_dict[partition_type])
     return matching_types
 
 
-def _valid_type_sequence(partition_types: List[Any]):
+def _valid_type_sequence(partition_types: List[Any], equivalet_types_dict=None):
```
[the remark this answers is not in the document]
Comment (leslie-fang-intel) [reply torch/ao/quantization/_pt2e/graph_utils.py:60]: Thanks and fix this and other spelling mistakes.

**Comment by leslie-fang-intel** [comment]
> I dont see any use of the added api in this stack. Can you add an example of why you need this change? If needed, we need to think of interface where it allows you to not just override but also extend the equivalent_types. e.g.
> 
> ```
> def get_equivalent_types()
> """
> defines a map from nn.Module equivalent nn.functional
> """
> ```

Hi @kimishpatel, thanks for the suggestions.

- Yean, PRs in this ghstack will not use it. So I move this PR to the last of this ghstack.
- We need this change, is because if we want to use ` Module Partition API` in PyTorch Extension Libraries such as [intel-extension-for-pytorch](https://github.com/intel/intel-extension-for-pytorch). We also need to define the customized `_EQUIVALENT_TYPES` in the Extension Libraries and override  `_EQUIVALENT_TYPES` hardcoded here. Do it sounds reasonable to you?
- If it's ok to this added api, what's your preferred interface?

**Review by jgong5 — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
> > * If it's ok to this added api, what's your preferred interface?
> 
> So what you have may work but I would like to have a way that allows it to also extend `_EQUIVALENT_TYPES`.
> 
> However, let me understand your usecase. Can you give me an example of how this will be used?

Thanks for the comment. Here is one of the example: 

- New nodes such as `mul`, `sigmoid` are not in the PyTorch `_EQUIVALENT_TYPES`. 
- We want to use `find_sequential_partitions` to match the pattern of `conv-sigmoid-mul` in PyTorch Extension Libraries. So we will use it as 
```
        fused_partitions = find_sequential_partitions(
            gm, [torch.nn.Conv2d, torch.sigmoid, torch.mul]
        )
```
However, `mul`, `sigmoid` are not in the PyTorch `_EQUIVALENT_TYPES`. So the match will fail to find the partitions. If we can use customized  `_EQUIVALENT_TYPES` should enable this use case.

**Comment by kimishpatel** [comment]
> > > * If it's ok to this added api, what's your preferred interface?
> > 
> > 
> > So what you have may work but I would like to have a way that allows it to also extend `_EQUIVALENT_TYPES`.
> > However, let me understand your usecase. Can you give me an example of how this will be used?
> 
> Thanks for the comment. Here is one of the example:
> 
> * New nodes such as `mul`, `sigmoid` are not in the PyTorch `_EQUIVALENT_TYPES`.
> * We want to use `find_sequential_partitions` to match the pattern of `conv-sigmoid-mul` in PyTorch Extension Libraries. So we will use it as
> 
> ```
>         fused_partitions = find_sequential_partitions(
>             gm, [torch.nn.Conv2d, torch.sigmoid, torch.mul]
>         )
> ```
> 
> However, `mul`, `sigmoid` are not in the PyTorch `_EQUIVALENT_TYPES`. So the match will fail to find the partitions. If we can use customized `_EQUIVALENT_TYPES` should enable this use case.

For that I would say you rather want to extend _EQUIVALENT_TYPE to include mul and sigmoid. Ideally we continue to increase coverage in _EQUIVALENT_TYPES such that it include wide range of pytorch ops, but in the interim when this is not the case, it is ok to have API that allows you to extend it. So I was gonna suggest to adding something like

```
native_types = get_equivalent_types()
native_types.extend(my_types)
# and then use this in the api you added.
```
However if my_types is just a subset of native pytorch types, I would like that to be just added to _EQUIVALENT_TYPES.

The other use case I can imagine is where you want to recognize your own nn.Module type. e.g. say you wrote `ConvAddSigmoidAndCustomPostProcess` and you want to find sequential partitions with `ConvAddSigmoidAndCustomPostProcess` + `ReLU`, then current method does not support that. But I also dont know if recognizing `ConvAddSigmoidAndCustomPostProcess` is supported in `get_source_partitions`. I will have to confirm.

So for now, I prefer you add types needed to `_EQUIVALENT_TYPES` directly, because it will benefit everyone.

**Comment by leslie-fang-intel** [comment]
Hi @kimishpatel, thanks for the suggestions. Agree that for such use case, 

- Firstly, we should add new ops (such as `mul`, `sigmoid`) in `_EQUIVALENT_TYPES` to increase its coverage and benefits every user.
- Also just as you mentioned, in the interim we may still need such interface to extend `_EQUIVALENT_TYPES` in PyTorch Extension Libraries. 
    - It can expedite the development efficiency.
    - It can also help validate the customized quantizer w/o change of PyTorch Code.

So I change the UX in the example of UT to

- Step 1: Use `graph_utils.get_equivalent_types` to get the `_EQUIVALENT_TYPES`  in Stock PyTorch.
- Step 2: Create `customized_equivalent_types` based on `_EQUIVALENT_TYPES`. 
  - Generally, user should append more equivalent ops into existing op set or add new op set into this list.
  - But in this UT, for purpose of easy maintain, I remove some ops from  `_EQUIVALENT_TYPES` to prove the customizable of `_EQUIVALENT_TYPES` .
- Step 3: Create `customized_equivalent_types_dict` by `_create_equivalent_types_dict(customized_equivalent_types)`
- Step 4: Invoke `find_sequential_partitions` with `customized_equivalent_types_dict`.

Please kindly let me know if this UX looks reasonable to you.

File: torch/ao/quantization/_pt2e/graph_utils.py [hunk 3]
```diff
@@ -20,15 +20,23 @@
 ]
 
 
-def _create_equivalent_types_dict():
+def get_equivalent_types() -> List[Set]:
+    return _EQUIVALENT_TYPES
+
+def _create_equivalent_types_dict(equivalent_types=None):
```
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/graph_utils.py:26]: You dont really need to override `_create_equivalent_types_dict`.

File: torch/ao/quantization/_pt2e/graph_utils.py [hunk 4]
```diff
@@ -20,15 +20,23 @@
 ]
 
 
-def _create_equivalent_types_dict():
+def get_equivalent_types() -> List[Set]:
```
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/graph_utils.py:23]: You only need this. and maybe `update_equivalent_types`

File: test/quantization/pt2e/test_graph_utils.py [hunk 5]
```diff
@@ -94,3 +98,39 @@ def forward(self, x):
...
+
+    @unittest.skipIf(IS_WINDOWS, "torch.compile is not supported on Windows")
+    def test_customized_equivalet_types_dict(self):
+        native_equivalent_types = get_equivalent_types()
+        customized_equivalent_types = native_equivalent_types
+
+        for equivalent_type_set in customized_equivalent_types:
+            if torch.nn.ReLU in equivalent_type_set:
+                equivalent_type_set.clear()
+                equivalent_type_set.add(torch.nn.ReLU)
```
Comment (kimishpatel) [inline test/quantization/pt2e/test_graph_utils.py:110]: Not sure what this test is doing really. You should check for some custom type.

**Review by kimishpatel — CHANGES_REQUESTED** [review]

**Comment by kimishpatel** [comment]
> 
> * Step 1: Use `graph_utils.get_equivalent_types` to get the `_EQUIVALENT_TYPES`  in Stock PyTorch.
> * Step 2: Create `customized_equivalent_types` based on `_EQUIVALENT_TYPES`.

Why do you need this? Can you not instead add `update_equivalent_types` that extends the type?
>   
>   * Generally, user should append more equivalent ops into existing op set or add new op set into this list.
>   * But in this UT, for purpose of easy maintain, I remove some ops from  `_EQUIVALENT_TYPES` to prove the customizable of `_EQUIVALENT_TYPES` .

For unit tests, why do you need to remove anything from equivalent_types? 


So at the end i expect just `get_equivalent_types`, `update_equivalent_types`. And this should serve the purpose.

The unit test where you remove and then add just doesnt really test the functionality you want. If you just want to test whether `update_equivalent_types` works, then just use another activation function like `leakyrelu` or something.

File: torch/ao/quantization/_pt2e/graph_utils.py (hunk 3 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/_pt2e/graph_utils.py]: Thanks for the comments and remove it.

File: torch/ao/quantization/_pt2e/graph_utils.py (hunk 4 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/_pt2e/graph_utils.py]: Great suggestion, it makes much less change to original code.

File: test/quantization/pt2e/test_graph_utils.py (hunk 5 above)
Comment (leslie-fang-intel) [reply test/quantization/pt2e/test_graph_utils.py]: Sure, changed.

**Comment by leslie-fang-intel** [comment]
@kimishpatel, thanks for the great suggestions. It makes much less change to original code. Please kindly help to take a look again.

File: torch/ao/quantization/_pt2e/graph_utils.py [hunk 6]
```diff
@@ -28,8 +28,21 @@ def _create_equivalent_types_dict():
...
+
+def get_equivalent_types() -> List[Set]:
+    return _EQUIVALENT_TYPES
+
+def update_equivalent_types_dict(customized_equivalent_types=None):
+    """Help function for user who wants to customize the _EQUIVALENT_TYPES and _EQUIVALENT_TYPES_DICT.
+    When customized_equivalent_types passes in,
+    re-generate _EQUIVALENT_TYPES and _EQUIVALENT_TYPES_DICT.
+    """
+    if customized_equivalent_types is not None:
```
Comment (kimishpatel) [inline torch/ao/quantization/_pt2e/graph_utils.py:41]: I would suggest you raise ValueError is update_equivalent_types==None. Because this function is explicitly for updating equivalent types

**Review by kimishpatel — APPROVED** [review]
Looks good. Thanks for the changes. Left a nit comment that you should fix before landing.

File: torch/ao/quantization/_pt2e/graph_utils.py (hunk 6 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/_pt2e/graph_utils.py:41]: Thanks for the suggestions and changed.
